pkg/volume/azure_dd/attacher.go

```diff
@@ -0,0 +1,289 @@
+}
+
+func (attacher *azureDiskAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMountPath string) error {
+	mounter := attacher.host.GetMounter()
+	notMnt, err := mounter.IsLikelyNotMountPoint(deviceMountPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(deviceMountPath, 0750); err != nil {
+				return err
+			}
+			notMnt = true
+		} else {
+			return err
+		}
+	}
+
+	volumeSource, err := getVolumeSource(spec)
+	if err != nil {
+		return err
+	}
+
+	options := []string{}
+	if spec.ReadOnly {
+		options = append(options, "ro")
+	}
+	if notMnt {
+		diskMounter := &mount.SafeFormatAndMount{Interface: mounter, Runner: exec.New()}
+		err = diskMounter.FormatAndMount(devicePath, deviceMountPath, *volumeSource.FSType, options)
+		if err != nil {
+			os.Remove(deviceMountPath)
+			return err
+		}
+	}
+	return nil
+}
+
+type azureDiskDetacher struct {
+	mounter mount.Interface
+	manager azureManager
+}
+
+var _ volume.Detacher = &azureDiskDetacher{}
+
+func (plugin *azureDataDiskPlugin) NewDetacher() (volume.Detacher, error) {
+	azure, err := getAzureDiskManager(plugin.host.GetCloudProvider())
+	if err != nil {
+		return nil, err
+	}
+
+	return &azureDiskDetacher{
+		mounter: plugin.host.GetMounter(),
+		manager: azure,
+	}, nil
+}
+
+func (detacher *azureDiskDetacher) Detach(dev string, hostName string) error {
+	if dev == "" {
+		return fmt.Errorf("invalid dev to detach: %q", dev)
+	}
+	instanceid, err := detacher.manager.InstanceID(hostName)
+	if err != nil {
+		return fmt.Errorf("failed to get azure instance id for host %q", hostName)
+	}
+	if ind := strings.LastIndex(instanceid, "/"); ind >= 0 {
+		instanceid = instanceid[(ind + 1):]
+	}
+	attachDetachMutex.LockKey(instanceid)
+	defer attachDetachMutex.UnlockKey(instanceid)
+
+	glog.V(4).Infof("detach %v from host %q", dev, instanceid)
+	err = detacher.manager.DetachDiskByName(dev, "", instanceid)
+	if err != nil {
+		glog.V(2).Infof("failed to detach azure disk %q, err %v", dev, err)
+	}
+
+	return err
+}
+
+func (detacher *azureDiskDetacher) WaitForDetach(devicePath string, timeout time.Duration) error {
+	ticker := time.NewTicker(checkSleepDuration)
+	defer ticker.Stop()
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			glog.V(4).Infof("Checking device %q is detached.", devicePath)
+			if pathExists, err := pathExists(devicePath); err != nil {
+				return fmt.Errorf("Error checking if device path exists: %v", err)
+			} else if !pathExists {
+				return nil
+			}
+		case <-timer.C:
+			return fmt.Errorf("Timeout reached; Device %v is still attached", devicePath)
+		}
+	}
+}
+
+func (detacher *azureDiskDetacher) UnmountDevice(deviceMountPath string) error {
+	volume := path.Base(deviceMountPath)
+	if err := unmountPDAndRemoveGlobalPath(deviceMountPath, detacher.mounter); err != nil {
+		glog.V(4).Infof("Error unmounting %q: %v", volume, err)
```

Comment:
adding @ciwang the unmount loop is from [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_executor.go#L896)

The cause here is that unlike GCE/AWS, Azure metadata service doesn't tell us the device path of mounted disk. So we use disk's LUN number as the identifier on detach controller and kubelet. The LUN number is then mapped to the device name on the host. As a result, probing the device identifier (i.e. LUN here) will fail.